lib/matplotlib/axes/_axes.py

```diff
@@ -2745,7 +2746,22 @@ def sign(x):
             if np.isnan(dat):
                 lbl = ''
 
-            annotation = self.annotate(fmt % value if lbl is None else lbl,
+            if lbl is None:
+                if callable(fmt):
+                    formatted_value = fmt(value)
+                elif isinstance(fmt, str):
+                    if fmt.count('{:') == fmt.count('}') == 1:
```

Comment:
This will fail to format `{{{:d}}}` which is a valid new-style format string that is not too unreasonable for a user to want.

Reply:
This is a classical case where *it's better to ask for forgiveness than for permission*.

It's probably a good idea to extract a function and put it into `matplotlib.cbook`, so that we can a) isolatedly test all kinds of input and b) reuse this later.

```
def _auto_format_str(fmt, value):
    """
    Apply *value* to the format string *fmt*.

    This works both with unnamed %-style formatting and
    unnamed {}-style formatting. %-style formatting has priority.
    If *fmt* is %-style formattable that will be used. Otherwise,
    {}-formatting is applied. Strings without formatting placeholders
    are passed through as is.

    Examples
    ---------
    >>> _auto_format_str('%.2f m', 0.2)
    '0.20 m'
    >>> _auto_format_str('{} m', 0.2)
    '0.2 m'
    >>> _auto_format_str('const', 0.2)
    'const'
    >>> _auto_format_str('%d or {}', 0.2)
    '0 or {}'
    """
    try:
        return = fmt % value
    except TypeError:
        return = fmt.format(value)
```

Reply:
I keep having an issue where the function tested on its own works, but the very same case passed into the function from `Axes.bar_label()` ends up not raising a `TypeError` when attempting %-style formatting on a {}-style format string like `{:,.0f}`. For this reason, I had to add an additional check that after attempting `fmt % value` that the result wasn't equivalent to `fmt`. I don't understand why that happens though.

Reply:
I'm probably wrong, but could `_preprocess_data` be messing with the calls stack in some way here 
https://github.com/matplotlib/matplotlib/blob/1d5221b2c5fe989fa6e7eb680839a1586cfa4d1e/lib/matplotlib/__init__.py#L1348